### PR TITLE
Use Dictionary for underlying cache of ResourceSet

### PR DESCRIPTION
**Pull request by marek-safar**
Reverts the revert and fixes regression reported as #44483

@tarekgh @jkotas

### CODE REVIEW

File: src/libraries/System.Private.CoreLib/src/System/Resources/ResourceSet.cs [hunk 1]
```diff
@@ -134,59 +124,46 @@ IEnumerator IEnumerable.GetEnumerator()
 
         private IDictionaryEnumerator GetEnumeratorHelper()
         {
-            Hashtable? copyOfTable = Table;  // Avoid a race with Dispose
-            if (copyOfTable == null)
-                throw new ObjectDisposedException(null, SR.ObjectDisposed_ResourceSet);
-            return copyOfTable.GetEnumerator();
+            // Avoid a race with Dispose
+            return (_table as IDictionary)?.GetEnumerator() ?? throw new ObjectDisposedException(null, SR.ObjectDisposed_ResourceSet);
```
Comment (jkotas) [inline src/libraries/System.Private.CoreLib/src/System/Resources/ResourceSet.cs:128]: I do not think that the cast to `IDictionary` here is necessary. `Dictionary.GetEnumerator()` and `IDictionary.GetEnumerator()` on `Dictionary` are returning the exact same enumerator.

Comment (jkotas) [reply src/libraries/System.Private.CoreLib/src/System/Resources/ResourceSet.cs:128]: Ah, ok. I see how this works. The enumerator is configured differently for IDictionary.

Comment (marek-safar) [reply src/libraries/System.Private.CoreLib/src/System/Resources/ResourceSet.cs:128]: Yep

File: src/libraries/System.Resources.ResourceManager/tests/ResourceSetTests.cs [hunk 2]
```diff
@@ -101,6 +102,87 @@ public void GetString()
...
+        {
+            var rs = new ResourceSet(new SimpleResourceReader());
+
+            var expected = new HashSet<object>() {
+                1, "String", "Int32"
+            };
+
+            foreach (DictionaryEntry entry in rs)
+            {
+                Assert.True (expected.Contains(entry.Key), entry.Key.ToString());
```
Comment (jkotas) [inline src/libraries/System.Resources.ResourceManager/tests/ResourceSetTests.cs:178]:
```suggestion
                Assert.Contains(entry.Key, expected);
```

File: src/libraries/System.Private.CoreLib/src/System/Resources/ResourceSet.cs (hunk 1 above)
Comment (jkotas) [inline src/libraries/System.Private.CoreLib/src/System/Resources/ResourceSet.cs:128]:
```suggestion
            IDictionary? copyOfTableAsIDictionary = _table;  // Avoid a race with Dispose
            if (copyOfTableAsIDictionary == null)
                throw new ObjectDisposedException(null, SR.ObjectDisposed_ResourceSet);
                
             // Use IDictionary.GetEnumerator() for backward compatibility. Callers expect the enumerator to return DictionaryEntry instances.
            return copyOfTableAsIDictionary.GetEnumerator();
```

**Review by jkotas — APPROVED** [review]
Thanks!

**Pull request merged**